<a href="https://colab.research.google.com/github/xkp88522/tainanoffice/blob/main/%22%E5%8F%B0%E5%8D%97%22%E5%9C%8B%E5%B0%8F%E8%AA%B2%E5%BE%8C%E7%A4%BE%E5%9C%98%E8%B3%87%E6%96%99_%E7%88%AC%E8%9F%B2_(%E5%B7%B2%E5%AE%8C%E6%88%90).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

##"台南"國小課後社團資料_爬蟲



規劃:
1. 先下載台南各地區國小名單(共三份)，並上傳至您的google雲端硬碟
(可自行建立名單，至少需要"學校名稱"及"網址")
2. 針對各國小的網頁原始碼做爬蟲(目前有3套方法，前2套只針對公佈欄的第一頁做爬取，第3套可以抓取完整的公布欄資訊)
3. 篩選出有符合"關鍵字"的結果(課後、社團、報名....)
4. 將資訊整理後儲存到"自己"的雲端硬碟，檔案名稱為「台南國小課後社團報名申請更新資訊_日期.csv」
5. 使用line權杖(可google查詢如何申請)做推播(群組或自己)


使用說明:
1. 請先在雲端硬碟另存一份新副本 (視窗左上角 > 檔案 > 在雲端硬碟中儲存副本)
2. 使用新副本依序執行每一個欄位的程式碼
3. 過程中會需要手動下載檔案並上傳到您的Google雲端硬碟
4. 執行到最下面的地方(最後一格)時，請"一定"要先修改成自己的Line權杖代碼
5. 重要訊息會以## **訊息內容** ##顯示，可以按照指示更改，其餘部分請勿隨意更改

備註:
1. 由於目前只會抓取每個學校的公布欄第一頁，若目標訊息被洗到第二頁的話，爬蟲是抓不到的，高峰期建議每天執行一次並記錄您關注的學校是否已發出社團申請訊息
2. 此程式碼目前只能分析台南地區的國小，如要分析其他縣市國小需要再修改測試

建立者: 謝昆陞 xkp88522@gmail.com 有問題請回報!

In [ ]:
## 安裝套件 ##
!pip install requests
!pip install beautifulsoup4
!pip install pandas
!pip install gdown
!pip install pytz

##載入台南各地區國小名單

In [ ]:
## 請先點選下方網址，下載整理好的名單(台南各區國小共有三份)，並上傳至您的google雲端硬碟 ##
file_url1 = 'https://drive.google.com/uc?id=1QViOWtYOIm9iF4VLxmnOaoTNORHuyLbY'  ## ctrl + 滑鼠左鍵 點擊下載檔案  ##
file_url2 = 'https://drive.google.com/uc?id=1H4qLUPO5cvp2CuDKx62SnwdQBwq6DnWy'  ## ctrl + 滑鼠左鍵 點擊下載檔案  ##
file_url3 = 'https://drive.google.com/uc?id=1e_mImsWUC2Ue7uZ7tbwiA0I8jsYQ2J21'  ## ctrl + 滑鼠左鍵 點擊下載檔案  ##

#連結到我的雲端硬碟
from google.colab import drive
import pandas as pd
drive.mount('/content/gdrive') # 此處需要登入google帳號
data1 = pd.read_csv("/content/gdrive/My Drive/台南各區國小_第一種抓取.csv",encoding="big5")
data2 = pd.read_csv("/content/gdrive/My Drive/台南各區國小_第二種抓取.csv",encoding="big5")
data3 = pd.read_csv("/content/gdrive/My Drive/台南各區國小_第三種抓取.csv",encoding="big5")

Mounted at /content/gdrive


In [ ]:
import pandas as pd

# 顯示資料
data =pd.concat([data1, data2, data3],axis=0,join="inner")
data

學校名稱       電話                        地址  \
0   大同國小  2157589      701台南市東區大福里25鄰榮譽街30號   
1   東光國小  2376534     701台南市東區東光里2鄰東光路1段39號   
2   德高國小  2681891    701台南市東區德高里17鄰崇善路1155號   
3   崇學國小  2689951       701台南市東區崇學里1鄰崇學路98號   
4   裕文國小  3317657     701台南市東區文聖里19鄰裕文路301號   
..   ...      ...                       ...   
34  東陽國小  7833322     726臺南市學甲區慈福里1鄰華宗路260號   
35  漚汪國小  7942012          725台南市將軍區長榮里5鄰四號   
36  新營國小  6322136      730臺南市新營區大宏里13鄰中正路4號   
37  新民國小  6562152  730臺南市新營區民生里18鄰公園路一段136號   
38  竹門國小  6851290       732台南市白河區竹門里10鄰113號   

                                                   網址 台南區域 抓取狀況  \
0   https://www.ttes.tn.edu.tw/modules/tadnews/ind...    東  第一種   
1   https://www.tkes.tn.edu.tw/modules/tadnews/ind...    東  第一種   
2   https://www.tkps.tn.edu.tw/modules/tadnews/ind...    東  第一種   
3   https://www.ches.tn.edu.tw/modules/tadnews/ind...    東  第一種   
4   https://www.ywes.tn.edu.tw/modules/tadnews/ind...    東  第一種   
..                                                ...  ...  ...   
34  http://www.dyps.tn.edu.tw/modules/tadnews/inde...   學甲  第三種   
35  http://www.owes.tn.edu.tw/modules/tadnews/inde...   將軍  第三種   
36  http://www.syps.tn.edu.tw/modules/tadnews/inde...   新營  第三種   
37  https://www.hmes.tn.edu.tw//modules/tadnews/in...   新營  第三種   
38  https://www.jmes.tn.edu.tw/modules/tadnews/ind...   白河  第三種   

                                     容器  \
0   'div', class_='news_page_container'   
1                                   NaN   
2                                   NaN   
3                                   NaN   
4                                   NaN   
..                                  ...   
34                                  NaN   
35                                  NaN   
36                                  NaN   
37                                  NaN   
38                                  NaN   

                                           標籤  
0   article.find('h4').find('a').text.strip()  
1                                         NaN  
2                                         NaN  
3                                         NaN  
4                                         NaN  
..                                        ...  
34                                        NaN  
35                                        NaN  
36                                        NaN  
37                                        NaN  
38                                        NaN  

[205 rows x 8 columns]

In [ ]:
df1a = data1["學校名稱"]
df1b = data1["網址"]
df1=pd.concat([df1a,df1b],axis=1,join="inner")

df2a = data2["學校名稱"]
df2b = data2["網址"]
df2=pd.concat([df2a,df2b],axis=1,join="inner")

df3a = data3["學校名稱"]
df3b = data3["網址"]
df3=pd.concat([df3a,df3b],axis=1,join="inner")

df3 # 只有學校名稱與網址的數據


學校名稱                                                 網址
0     勝利國小  https://www.slps.tn.edu.tw/modules/tadnews/ind...
1     崇明國小  https://www.cmes.tn.edu.tw/modules/tadnews/ind...
2     志開實小  https://www.zkes.tn.edu.tw/modules/tadnews/ind...
3     文元國小  https://www.wyes.tn.edu.tw/modules/tadnews/ind...
4     新南國小  https://www.hnps.tn.edu.tw/modules/tadnews/ind...
5     西門實小  https://www.hmps.tn.edu.tw/modules/tadnews/ind...
6     億載國小           https://www.yzes.tn.edu.tw/news_list.asp
7     和順國小  https://www.hses.tn.edu.tw/modules/tadnews/ind...
8     海東國小  https://www.htps.tn.edu.tw/modules/tadnews/ind...
9     大灣國小  https://www.dwps.tn.edu.tw/modules/tadnews/ind...
10    三村國小  https://www.stes.tn.edu.tw/modules/tadnews/ind...
11  永康復興國小  https://www.fses.tn.edu.tw/modules/tadnews/ind...
12    崑山國小  https://www.kses.tn.edu.tw/modules/tadnews/ind...
13    仁德國小           http://120.116.50.2/school/pub/index.php
14    那拔國小  https://www.nbes.tn.edu.tw/modules/tadnews/ind...
15    正新國小  https://www.jses.tn.edu.tw/modules/tadnews/ind...
16    大社國小  https://www.dsps.tn.edu.tw/modules/tadnews/ind...
17    安定國小  https://www.adps.tn.edu.tw/modules/tadnews/ind...
18    港東國小  https://www.gdes.tn.edu.tw/modules/tadnews/ind...
19    後營國小  https://www.hyes.tn.edu.tw/modules/tadnews/ind...
20    松林國小  https://www.soles.tn.edu.tw/modules/tadnews/in...
21    七股國小  https://www.cgps.tn.edu.tw/modules/tadnews/ind...
22    三股國小  https://www.sgps.tn.edu.tw/modules/tadnews/ind...
23    龍山國小  https://www.lses.tn.edu.tw/modules/tadnews/ind...
24    佳里國小  https://www.jles.tn.edu.tw/modules/tadnews/ind...
25    北勢國小  https://www.bses.tn.edu.tw/modules/tadnews/ind...
26    陽明國小  https://www.ymes.tn.edu.tw/modules/tadnews/ind...
27    小新國小  https://www.ssees.tn.edu.tw/modules/tadnews/in...
28    山上國小  http://www.sases.tn.edu.tw/modules/tadnews/ind...
29    光榮實小  http://www.gres.tn.edu.tw/modules/tadnews/inde...
30    南化國小  http://www.nhes.tn.edu.tw/modules/tadnews/inde...
31    二溪國小  https://www.eses.tn.edu.tw/modules/tadnews/ind...
32    隆田國小  http://www.ltes.tn.edu.tw/modules/tadnews/inde...
33    官田國小  http://www.gtes.tn.edu.tw//modules/tadnews/ind...
34    東陽國小  http://www.dyps.tn.edu.tw/modules/tadnews/inde...
35    漚汪國小  http://www.owes.tn.edu.tw/modules/tadnews/inde...
36    新營國小  http://www.syps.tn.edu.tw/modules/tadnews/inde...
37    新民國小  https://www.hmes.tn.edu.tw//modules/tadnews/in...
38    竹門國小  https://www.jmes.tn.edu.tw/modules/tadnews/ind...

## 針對名單中的所有學校做爬蟲  
目前台南地區共205校，總約花費15-20分鐘

注意: 爬蟲過程中可能有抓取失敗，請再自行另外確認!   (目前只有仁德國小還抓不到)

In [ ]:
# 第一種抓取 容器: div , 標籤: h4 中的 a
import requests
from bs4 import BeautifulSoup
import pandas as pd

def crawl_website(url, school_name):
    df = pd.DataFrame(columns=['國小', '標題', '來源網址'])

    full_url = url

    response = requests.get(full_url)
    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')

        # 找到所有文章容器
        articles = soup.find_all('div', class_='news_page_container')

        for article in articles:
            title = article.find('h4').find('a').text.strip()

            # 過濾只保留包含"社團"、"課後"、"照顧"和"才藝"的標題  ## 可更改 ##
            if '社團' in title or '課後' in title or '照顧' in title or '才藝' in title:
                df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)

    else:
        print(f"Error: {response.status_code} for {full_url}")

    return df


original_data = df1   # pd.DataFrame(data)

# 逐一爬蟲每一筆資料
result_dfs = []
for index, row in original_data.iterrows():
    school_name = row['學校名稱']
    url = row['網址']

    # 呼叫爬取函數
    result_df = crawl_website(url, school_name)

    # 如果結果不為空，輸出 DataFrame
    if not result_df.empty:
        result_dfs.append(result_df)

# 合併所有結果
final_result_df1 = pd.concat(result_dfs, ignore_index=True)

# 顯示最終結果
print(final_result_df1)

<ipython-input-5-e5d153c9ed0e>:23: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)
<ipython-input-5-e5d153c9ed0e>:23: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)
<ipython-input-5-e5d153c9ed0e>:23: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)
<ipython-input-5-e5d153c9ed0e>:23: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_u

      國小                                                 標題  \
0   東光國小                        公告112學年度寒假及第二學期課後社團暨照顧班錄取名單   
1   裕文國小                           公布112學年度第2學期課後社團第二階段錄取名單   
2   成功國小                               112學年度第二學期課後才藝社團線上報名   
3   安平國小  112學年度下學期課後照顧班於113年2月16日(星期五)開始上課；2/17(六)調整上班上...   
4   海佃國小                           112學年度第二學期的課後社團報名至2/20截止   
5   永康國小                                112學年度第二學期課後社團-報名通知   
6   新市國小  轉知-臺南市113年度「校園病媒蚊蟲在地生活深耕-防蚊實驗社團暨創意宣導活動/競賽」推動計畫...   
7   大山國小  檢送本校推廣教育「113年度兒童課後照顧服務人員職前 訓練課程」平日班招生訊息，敬請轉知並鼓...   
8   北寮國小                                    龍耀南化慶豐年 學子展多元才藝   
9   仁光國小                   轉知社團法人大享食育協會製作「解謎豆子－雜糧知識問答」活動訊息。   
10  南梓實小                  112學年度第二學期「課後社團」錄取名單 (2/26起，正式上課)   

                                                 來源網址  
0   https://www.tkes.tn.edu.tw/modules/tadnews/ind...  
1   https://www.ywes.tn.edu.tw/modules/tadnews/ind...  
2   https://www.ckes.tn.edu.tw/modules/tadnews/ind...  
3   https://www.app

In [ ]:
# 第二種抓取 容器: table ， 標籤: td 中 的 a
import requests
from bs4 import BeautifulSoup
import pandas as pd

def crawl_website(url, school_name):
    df = pd.DataFrame(columns=['國小', '標題', '來源網址'])

    full_url = url

    response = requests.get(full_url)
    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')

       # 找到所有文章容器
        articles = soup.select('table.table.table-striped.table-hover.table-shadow tr')

        for article in articles:
            td_tag = article.find('td')
            if td_tag:
                date = td_tag.text.strip().split('\n')[0]

                badge = article.find('a', class_='badge')
                category = badge.text if badge else '其他'

                a_tag = td_tag.find('a')
                if a_tag:
                    title = a_tag.text.strip()
                    source_url = a_tag['href']

                    # 過濾只保留包含特定關鍵字的標題，如: '社團', '課後', '照顧', '才藝' ## 可更改 ##
                    keywords = ['社團', '課後', '照顧', '才藝']
                    if any(keyword in title for keyword in keywords):
                        df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)
    else:
        print(f"Error: {response.status_code} for {full_url}")

    return df


original_data = df2   # pd.DataFrame(data)

# 逐一爬蟲每一筆資料
result_dfs = []
for index, row in original_data.iterrows():
    school_name = row['學校名稱']
    url = row['網址']

    # 呼叫爬取函數
    result_df = crawl_website(url, school_name)

    # 如果結果不為空，輸出 DataFrame
    if not result_df.empty:
        result_dfs.append(result_df)

# 合併所有結果
final_result_df2 = pd.concat(result_dfs, ignore_index=True)

# 顯示最終結果
print(final_result_df2)


<ipython-input-6-15cbee9b0b8b>:34: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)
<ipython-input-6-15cbee9b0b8b>:34: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)
<ipython-input-6-15cbee9b0b8b>:34: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_url}, ignore_index=True)
<ipython-input-6-15cbee9b0b8b>:34: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': title, '來源網址': full_u

         國小                                  標題  \
0      省躬國小  社團法人中華國際身心障礙者生命教育推廣協會辦理【環保愛台灣】繪畫比賽   
1  九份子國(中)小             112學年度第2學期課後社團錄取上課名單及地點   
2      永信國小     臺南市永康區永信國民小學附設幼兒112年第一學期延長照顧調查表   
3    善化大同國小                   112年度兒童課後照顧服務人員培訓   

                                                來源網址  
0  https://www.sges.tn.edu.tw/modules/tadnews/ind...  
1  https://www.jfzjps.tn.edu.tw/modules/tadnews/i...  
2  https://www.yuses.tn.edu.tw/modules/tadnews/in...  
3  https://www.dtes.tn.edu.tw/modules/tadnews/ind...  


In [ ]:
# 第三種抓取 容器: table  標籤: td 中 的 a
import requests
from bs4 import BeautifulSoup
import pandas as pd

def crawl_website(url, school_name):
    df = pd.DataFrame(columns=['國小', '標題', '來源網址'])

    full_url = url

    response = requests.get(full_url)
    if response.status_code == 200:
        soup = BeautifulSoup(response.text, 'html.parser')

        # 找到所有文章容器
        articles = soup.select('table.table.table-striped.table-hover.table-shadow tr')

        for article in articles:
            td_tag = article.find('td')
            if td_tag:
                # 使用 .stripped_strings 取得所有文本
                td_text = ' '.join(td_tag.stripped_strings)

                badge = article.find('a', class_='badge')
                category = badge.text if badge else '其他'

                a_tag = td_tag.find('a')
                if a_tag:
                    title = a_tag.text.strip()
                    source_url = a_tag['href']

                    # 過濾只保留包含特定關鍵字的標題
                   # if '社團' in td_text or '課後' in td_text or '才藝' in td_text or '照顧' in td_text:
                    if '社團' in td_text or '課後' in td_text or '才藝' in td_text or '照顧' in td_text:
                        df = df.append({'國小': school_name, '標題': td_text, '來源網址': full_url}, ignore_index=True)
    else:
        print(f"Error: {response.status_code} for {full_url}")

    return df


original_data = df3   # pd.DataFrame(data)

# 逐一爬蟲每一筆資料
result_dfs = []
for index, row in original_data.iterrows():
    school_name = row['學校名稱']
    url = row['網址']

    # 呼叫爬取函數
    result_df = crawl_website(url, school_name)

    # 如果結果不為空，輸出 DataFrame
    if not result_df.empty:
        result_dfs.append(result_df)

# 合併所有結果
final_result_df3 = pd.concat(result_dfs, ignore_index=True)

# 顯示最終結果
print(final_result_df3)


<ipython-input-7-c115c27f0aae>:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': td_text, '來源網址': full_url}, ignore_index=True)
<ipython-input-7-c115c27f0aae>:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': td_text, '來源網址': full_url}, ignore_index=True)
<ipython-input-7-c115c27f0aae>:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': td_text, '來源網址': full_url}, ignore_index=True)
<ipython-input-7-c115c27f0aae>:35: FutureWarning: The frame.append method is deprecated and will be removed from pandas in a future version. Use pandas.concat instead.
  df = df.append({'國小': school_name, '標題': td_text, '來源網址'

      國小                                                 標題  \
0   崇明國小  2024-02-19 公告 112學年度第二學期課後社團報名時間2/19(一)08：00〜2...   
1   和順國小  2024-02-07 輔導室 轉知佛光山台南福國寺「好苗子幸福學堂」課後學習輔導報名訊息 (...   
2   海東國小  2024-02-16 公告 海東國小112學年度下學期才藝社團簡介 ( 學生活動組 / 68...   
3   海東國小  2024-02-19 公告 112學年第二學期 親子機器人程式設計課後社團上課日 ( 資訊組...   
4   大灣國小                2024-02-16 公告 112學年度第二學期課後學藝各社團上課地點   
5   大灣國小  2023-12-23 公告 臺南市永康區大灣國小附設幼兒園113年度寒假延長照顧服務特教助理...   
6   大灣國小  2023-11-20 狂賀 恭喜本校特教班參加臺南市112學年度特殊教育學生才藝比賽榮獲團體...   
7   三村國小  2024-02-15 公告 112-2_課後照顧班 2/16(五)開始上課，2/17(六)也...   
8   崑山國小  2024-02-19 超重要 公告112學年度第2學期課後才藝開課班級及上課地點. ( 黃惠...   
9   崑山國小  2024-02-15 超重要 崑山國小112學年度第2學期課後才藝社團線上報名.113年2月...   
10  陽明國小  2024-01-29 幼兒園 陽明國小附幼 112學年度寒假延長照顧點心與午餐表 ( 陳怡靖...   
11  新民國小  2024-02-19 學務處 新民112下學期 "課後社團" 開課日公告(2/19修訂週四舞...   

                                                 來源網址  
0   https://www.cmes.tn.edu.tw/modules/tadnews/ind...  
1   https://www.hses.tn.edu.tw/modules/tadnews/ind...  
2   https://

In [ ]:
# 合併三筆搜尋不同標籤的結果
final_result_df = pd.concat([ final_result_df1, final_result_df2, final_result_df3 ],axis=0,join="inner", ignore_index=True)
final_result_df # 只有學校名稱與網址的數據

國小                                                 標題  \
0       東光國小                        公告112學年度寒假及第二學期課後社團暨照顧班錄取名單   
1       裕文國小                           公布112學年度第2學期課後社團第二階段錄取名單   
2       成功國小                               112學年度第二學期課後才藝社團線上報名   
3       安平國小  112學年度下學期課後照顧班於113年2月16日(星期五)開始上課；2/17(六)調整上班上...   
4       海佃國小                           112學年度第二學期的課後社團報名至2/20截止   
5       永康國小                                112學年度第二學期課後社團-報名通知   
6       新市國小  轉知-臺南市113年度「校園病媒蚊蟲在地生活深耕-防蚊實驗社團暨創意宣導活動/競賽」推動計畫...   
7       大山國小  檢送本校推廣教育「113年度兒童課後照顧服務人員職前 訓練課程」平日班招生訊息，敬請轉知並鼓...   
8       北寮國小                                    龍耀南化慶豐年 學子展多元才藝   
9       仁光國小                   轉知社團法人大享食育協會製作「解謎豆子－雜糧知識問答」活動訊息。   
10      南梓實小                  112學年度第二學期「課後社團」錄取名單 (2/26起，正式上課)   
11      省躬國小                 社團法人中華國際身心障礙者生命教育推廣協會辦理【環保愛台灣】繪畫比賽   
12  九份子國(中)小                            112學年度第2學期課後社團錄取上課名單及地點   
13      永信國小                    臺南市永康區永信國民小學附設幼兒112年第一學期延長照顧調查表   
14    善化大同國小                                  112年度兒童課後照顧服務人員培訓   
15      崇明國小  2024-02-19 公告 112學年度第二學期課後社團報名時間2/19(一)08：00〜2...   
16      和順國小  2024-02-07 輔導室 轉知佛光山台南福國寺「好苗子幸福學堂」課後學習輔導報名訊息 (...   
17      海東國小  2024-02-16 公告 海東國小112學年度下學期才藝社團簡介 ( 學生活動組 / 68...   
18      海東國小  2024-02-19 公告 112學年第二學期 親子機器人程式設計課後社團上課日 ( 資訊組...   
19      大灣國小                2024-02-16 公告 112學年度第二學期課後學藝各社團上課地點   
20      大灣國小  2023-12-23 公告 臺南市永康區大灣國小附設幼兒園113年度寒假延長照顧服務特教助理...   
21      大灣國小  2023-11-20 狂賀 恭喜本校特教班參加臺南市112學年度特殊教育學生才藝比賽榮獲團體...   
22      三村國小  2024-02-15 公告 112-2_課後照顧班 2/16(五)開始上課，2/17(六)也...   
23      崑山國小  2024-02-19 超重要 公告112學年度第2學期課後才藝開課班級及上課地點. ( 黃惠...   
24      崑山國小  2024-02-15 超重要 崑山國小112學年度第2學期課後才藝社團線上報名.113年2月...   
25      陽明國小  2024-01-29 幼兒園 陽明國小附幼 112學年度寒假延長照顧點心與午餐表 ( 陳怡靖...   
26      新民國小  2024-02-19 學務處 新民112下學期 "課後社團" 開課日公告(2/19修訂週四舞...   

                                                 來源網址  
0   https://www.tkes.tn.edu.tw/modules/tadnews/ind...  
1   https://www.ywes.tn.edu.tw/modules/tadnews/ind...  
2   https://www.ckes.tn.edu.tw/modules/tadnews/ind...  
3   https://www.apps.tn.edu.tw/modules/tadnews/ind...  
4   https://www.hdes.tn.edu.tw/modules/tadnews/ind...  
5   https://www.ykes.tn.edu.tw/modules/tadnews/ind...  
6   https://www.sinses.tn.edu.tw/modules/tadnews/i...  
7   https://www.dses.tn.edu.tw/modules/tadnews/ind...  
8   https://www.bles.tn.edu.tw/modules/tadnews/ind...  
9   http://www.rges.tn.edu.tw/modules/tadnews/inde...  
10  http://www.ntes.tn.edu.tw/modules/tadnews/inde...  
11  https://www.sges.tn.edu.tw/modules/tadnews/ind...  
12  https://www.jfzjps.tn.edu.tw/modules/tadnews/i...  
13  https://www.yuses.tn.edu.tw/modules/tadnews/in...  
14  https://www.dtes.tn.edu.tw/modules/tadnews/ind...  
15  https://www.cmes.tn.edu.tw/modules/tadnews/ind...  
16  https://www.hses.tn.edu.tw/modules/tadnews/ind...  
17  https://www.htps.tn.edu.tw/modules/tadnews/ind...  
18  https://www.htps.tn.edu.tw/modules/tadnews/ind...  
19  https://www.dwps.tn.edu.tw/modules/tadnews/ind...  
20  https://www.dwps.tn.edu.tw/modules/tadnews/ind...  
21  https://www.dwps.tn.edu.tw/modules/tadnews/ind...  
22  https://www.stes.tn.edu.tw/modules/tadnews/ind...  
23  https://www.kses.tn.edu.tw/modules/tadnews/ind...  
24  https://www.kses.tn.edu.tw/modules/tadnews/ind...  
25  https://www.ymes.tn.edu.tw/modules/tadnews/ind...  
26  https://www.hmes.tn.edu.tw//modules/tadnews/in...

In [ ]:
# 將 DataFrame 轉換成 Markdown 格式的表格
markdown_table = final_result_df.to_markdown()

# 顯示 Markdown 表格
print(markdown_table)


|    | 國小           | 標題                                                                                                                                                                      | 來源網址                                               |
|---:|:---------------|:--------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------|
|  0 | 東光國小       | 公告112學年度寒假及第二學期課後社團暨照顧班錄取名單                                                                                                                       | https://www.tkes.tn.edu.tw/modules/tadnews/index.php   |
|  1 | 裕文國小       | 公布112學年度第2學期課後社團第二階段錄取名單                                                                                                                              | https://www.ywes.tn.edu.tw/modules/tadnews/index.php   |
|  2 | 成功國小       | 112學年度第二學期課後才藝社團線上報名        

## 篩選有"課後社團報名/申請"的國小


In [ ]:
# 挑選出包含"開課"、"申請"、"報名簡章"、"報名資料"、"報名時間"的標題

import pandas as pd

# 使用正規表達式進行多個關鍵字的匹配，並排除包含"結果"的標題
keywords = ['開課', '申請', '報名', '報名簡章', '報名資料', '報名時間']  ## 搜尋的關鍵字，可修改 ##
contains_keywords = '|'.join(keywords)
excludes_empty = ~final_result_df['標題'].str.contains('照顧')  ## 排除的關鍵字，可修改 ##

# 使用 str.contains() 進行篩選
contains_desired_keywords = final_result_df['標題'].str.contains(contains_keywords)

# 選取符合條件的列
pick_result_df = final_result_df[contains_desired_keywords & excludes_empty]

# 顯示結果
print(pick_result_df)


      國小                                                 標題  \
2   成功國小                               112學年度第二學期課後才藝社團線上報名   
4   海佃國小                           112學年度第二學期的課後社團報名至2/20截止   
5   永康國小                                112學年度第二學期課後社團-報名通知   
15  崇明國小  2024-02-19 公告 112學年度第二學期課後社團報名時間2/19(一)08：00〜2...   
16  和順國小  2024-02-07 輔導室 轉知佛光山台南福國寺「好苗子幸福學堂」課後學習輔導報名訊息 (...   
23  崑山國小  2024-02-19 超重要 公告112學年度第2學期課後才藝開課班級及上課地點. ( 黃惠...   
24  崑山國小  2024-02-15 超重要 崑山國小112學年度第2學期課後才藝社團線上報名.113年2月...   
26  新民國小  2024-02-19 學務處 新民112下學期 "課後社團" 開課日公告(2/19修訂週四舞...   

                                                 來源網址  
2   https://www.ckes.tn.edu.tw/modules/tadnews/ind...  
4   https://www.hdes.tn.edu.tw/modules/tadnews/ind...  
5   https://www.ykes.tn.edu.tw/modules/tadnews/ind...  
15  https://www.cmes.tn.edu.tw/modules/tadnews/ind...  
16  https://www.hses.tn.edu.tw/modules/tadnews/ind...  
23  https://www.kses.tn.edu.tw/modules/tadnews/ind...  
24  https://www.kses.tn.edu.tw/modules/t

##上傳資料到Google雲端硬碟

In [ ]:
from google.colab import drive
import pandas as pd
from datetime import datetime

# 連結到我的雲端硬碟
drive.mount('/content/gdrive') ## 此處需要登入 Google 帳號  ##

# 取得當前日期
current_date = datetime.now().strftime("%Y-%m-%d")

# 資料儲存到我的雲端硬碟，檔名包含日期
file_name = f"/content/gdrive/My Drive/台南國小課後社團報名申請更新資訊_{current_date}.csv"
pick_result_df.to_csv(file_name, encoding="utf-8-sig")


print(f"檔案已儲存為: {file_name}")

In [ ]:
# 整理要推播到LINE_Notify的訊息內容
# 在輸出時去除整列「標題」和「來源網址」，並使用 drop_duplicates 移除內容相同的整列

# 去除 '標題' 和 '來源網址' 欄位
send_pick_result_df = pick_result_df.drop(['標題', '來源網址'], axis=1)

# 去除重複的整列
send_pick_result_df = send_pick_result_df.drop_duplicates()

# 以純文字格式顯示結果，不包含標頭和 index
result_text = send_pick_result_df.to_string(header=False, index=False)

# 顯示結果
print(result_text)


##將爬蟲結果傳送到LINE_Notify

In [ ]:
import requests
from datetime import datetime
import pytz

# LINE Notify 的權杖(token)
token = "Xms2fkdjTNLzBibyrH6miYKpjVs6g5OcuovwIpglK5n"  ## 填入權杖代碼，請修改成自己的 ##

# 傳送 LINE Notify 訊息
def send_line_notify(token, msg):
    headers = {
        "Authorization": "Bearer " + token,
        "Content-Type": "application/x-www-form-urlencoded"
    }

    # 設定時區為 Asia/Taipei (台灣)
    taipei_timezone = pytz.timezone('Asia/Taipei')

    # 取得當前時間，並轉換為指定時區
    current_time_utc = datetime.utcnow()
    current_time_taipei = current_time_utc.replace(tzinfo=pytz.utc).astimezone(taipei_timezone)

    # 格式化時間為字串
    current_time_str = current_time_taipei.strftime("%Y-%m-%d %H:%M:%S")

    data = {
        'message': f"{msg}\n\n訊息發送時間：{current_time_str}",  # 將時間加入訊息中
        "stickerPackageId": 6362,  # 貼圖包ID
        "stickerId": 11087933  # 貼圖ID
    }

    r = requests.post("https://notify-api.line.me/api/notify", headers=headers, data=data)

# 呼叫傳送 LINE Notify 函數
send_line_notify(token, "\n 最新資訊已更新至您的雲端 \n " + "https://drive.google.com/drive" + "\n 名單如下: \n" + result_text)

print("訊息傳送完成")